src/transforms/kv_parser.rs

```diff
@@ -0,0 +1,404 @@
+use string_cache::DefaultAtom as Atom;
+
+#[derive(Debug, Default, Derivative, Deserialize, Serialize)]
+#[serde(default, deny_unknown_fields)]
+pub struct KeyValueConfig {
+    #[derivative(Default(value = "true"))]
+    pub drop_field: bool,
+    pub field: Option<Atom>,
+    #[derivative(Default(value = " "))]
+    pub field_split: String,
+    #[derivative(Default(value = "true"))]
+    pub overwrite_target: bool,
+    pub separator: String,
+    pub target_field: Option<Atom>,
+    pub trim_key: Option<String>,
+    pub trim_value: Option<String>,
+    pub types: HashMap<Atom, String>,
+}
+
+inventory::submit! {
+    TransformDescription::new::<KeyValueConfig>("kv_parser")
+}
+
+#[typetag::serde(name = "kv_parser")]
+impl TransformConfig for KeyValueConfig {
+    fn build(&self, _cx: TransformContext) -> crate::Result<Box<dyn Transform>> {
+        let field = self
+            .field
+            .as_ref()
+            .unwrap_or(&event::log_schema().message_key());
+
+        let conversions = parse_conversion_map(&self.types)?;
+
+        Ok(Box::new(KeyValue {
+            conversions,
+            drop_field: self.drop_field,
+            field: field.clone(),
+            field_split: self.field_split.clone(),
+            overwrite_target: self.overwrite_target,
+            separator: self.separator.clone(),
+            target_field: self.target_field.clone(),
+            trim_key: self.trim_key.clone(),
+            trim_value: self.trim_value.clone(),
+        }))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn transform_type(&self) -> &'static str {
+        "kv_parser"
+    }
+}
+
+pub struct KeyValue {
+    conversions: HashMap<Atom, Conversion>,
+    drop_field: bool,
+    field: Atom,
+    field_split: String,
+    overwrite_target: bool,
+    separator: String,
+    target_field: Option<Atom>,
+    trim_key: Option<String>,
+    trim_value: Option<String>,
+}
+
+fn kv_parser(
+    pair: String,
+    field_split: &String,
+    trim_key: &Option<String>,
+    trim_value: &Option<String>,
+) -> Option<(Atom, String)> {
+    let pair = pair.trim();
+    let mut field;
+
+    if field_split.is_empty() {
+        let mut kv_pair = pair.split_whitespace();
+        let key = kv_pair.nth(0)?;
+        let val = kv_pair.nth(1)?;
+        let count = kv_pair.count();
+
+        if count < 2 {
+            return None;
+        } else if count > 2 {
+            error!(
+                message = "KV parser expected 2 values, but got {count}",
+                count = count,
+                rate_limit_secs = 30
+            );
+            return None;
+        }
+
+        field = [key, val];
+    } else {
+        let split_index = pair.find(field_split).unwrap_or(0);
+        let (key, val) = pair.split_at(split_index);
+        let key = key.trim();
+        let val = val.trim_start_matches(field_split).trim();
```

Comment:
```suggestion
        let val = pair[split_index + len(field_split)..].trim();
```
It's too bad Rust's strings don't have a built-in split that handles this internally.

Reply:
Yeah, I couldn't find anything. Kind of sad to hear this is the best we can do with the standard lib.

I'm guessing this is more performant than `trim`?

Reply:
I'm not sure what you're asking about `trim`, since it's still included. However note that the `trim*` functions remove _all matches_ rather than just the first, so if you want to only remove one they are not appropriate.

Strings are indexed at byte offsets not character indices, so your second suggestion is correct. I still sometimes think `len(X)` from long use of Python.

The use of `split_at` with a separator isn't particularly useful, as it will always require post-processing to remove the separator.

Reply:
Sorry, I was talking about `trim_start_matches` and the diff here. As far as I could tell, they both do the same thing? If that's not the case, can you point me to some docs or elaborate? If it is, what are the other reasons for the suggestion?

Reply:
`trim_start_matches` may strip multiple matches. So, if the separator is `=` and the string is `key===value` then `val` will contain `value` with `trim_start_matches`, but `==value` with the manual split. This is an important difference.